### PR TITLE
AutoMLSearch: hold reference to training data

### PR DESCRIPTION
**Pull request by dsherry**
**Changes**
* Update `AutoMLSearch` constructor to take training data as `X_train` and `y_train`. Update `search` and `add_to_leaderboard` to no longer require data
* Remove the woodwork warnings when pandas dtypes are provided (@angela97lin heads up this will cause mild conflicts with #1555 . I can try to back it out of this PR if you'd like, but it felt easier in the moment to just remove here heh)
* Update `split_data` helper to accept `problem_type`, `problem_configuration`, `test_size`, `shuffle`, and to default `random_state` to 0
* Rename the "Testing" split (during CV) in the automl log output to "Validation"
* Move `evalml/automl/data_splitters/` to `evalml/preprocessing/data_splitters/` to avoid circular import
* Add problem type utils: `is_regression`, `is_classification`, `is_timeseries`, `is_binary`, `is_multiclass`
* Update documentation

### CODE REVIEW

File: evalml/automl/automl_search.py [hunk 1]
```diff
@@ -402,24 +401,14 @@ def search(self, X, y, data_checks="auto", show_iteration_plot=True):
...
-
-        text_column_vals = X.select('natural_language')
+        text_column_vals = self.X_train.select('natural_language')
         text_columns = list(text_column_vals.to_dataframe().columns)
         if len(text_columns) == 0:
             text_columns = None
 
-        if not isinstance(y, ww.DataColumn):
-            logger.warning("`y` passed was not a DataColumn. EvalML will try to convert the input as a Woodwork DataTable and types will be inferred. To control this behavior, please pass in a Woodwork DataTable instead.")
-            y = _convert_to_woodwork_structure(y)
```
Comment (dsherry) [inline evalml/automl/automl_search.py:417]: I removed these warnings because we were planning on doing so (#1555) and because it was easier than fixing the failing test coverage for them 😅

File: evalml/automl/automl_search.py [hunk 2]
```diff
@@ -614,29 +599,26 @@ def _get_mean_cv_scores_for_all_objectives(cv_data):
                     scores[field] += value
         return {objective_name: float(score) / n_folds for objective_name, score in scores.items()}
 
-    def _compute_cv_scores(self, pipeline, X, y):
+    def _compute_cv_scores(self, pipeline):
```
Comment (dsherry) [inline evalml/automl/automl_search.py:602]: It would be great to get more eyes on the changes here. I think I did it right but can't hurt to check.

File: evalml/automl/automl_search.py [hunk 3]
```diff
@@ -683,7 +667,7 @@ def _compute_cv_scores(self, pipeline, X, y):
             ordered_scores.update({self.objective.name: score})
             ordered_scores.update(scores)
             ordered_scores.update({"# Training": y_train.shape[0]})
-            ordered_scores.update({"# Testing": y_test.shape[0]})
+            ordered_scores.update({"# Validation": y_valid.shape[0]})
```
Comment (dsherry) [inline evalml/automl/automl_search.py:670]: I renamed this field because I think "validation" is a better description

File: evalml/preprocessing/data_splitters/time_series_split.py [hunk 4]
```diff
@@ -21,6 +21,8 @@ def __init__(self, max_delay=0, gap=0, n_splits=3):
                 of rows of the current split to avoid "throwing out" more data than in necessary.
             gap (int): Gap used in time series problem. Time series pipelines shift the target variable by gap rows
                 since we are interested in
+            n_splits (int): number of data splits to make.
+            test_size (float): What percentage of data points should be included in the test set. Defaults to None.
```
Comment (dsherry) [inline evalml/preprocessing/data_splitters/time_series_split.py:25]: Oops, I will remove `test_size`, left over

File: evalml/preprocessing/data_splitters/training_validation_split.py [hunk 5]
```diff
@@ -43,4 +43,4 @@ def split(self, X, y=None):
                 list: indices to split data into training and test set
         """
         train, test = train_test_split(np.arange(X.shape[0]), test_size=self.test_size, train_size=self.train_size, shuffle=self.shuffle, stratify=self.stratify, random_state=self.random_state)
-        return [(train, test)]
+        return iter([(train, test)])
```
Comment (dsherry) [inline evalml/preprocessing/data_splitters/training_validation_split.py:46]: I believe this was necessary because `BaseCrossValidator` subclasses are expected to return iterators. I had to update the tests to call `list(...)` 🤷‍♂️

File: evalml/preprocessing/utils.py [hunk 6]
```diff
@@ -55,16 +63,15 @@ def split_data(X, y, regression=False, test_size=.2, random_state=None):
...
-                                 random_state=random_state)
-    else:
-        CV_method = StratifiedShuffleSplit(
-            n_splits=1,
-            test_size=test_size,
-            random_state=random_state)
-    train, test = next(CV_method.split(X.to_dataframe(), y.to_series()))
+    data_splitter = None
+    if is_time_series(problem_type):
+        data_splitter = TrainingValidationSplit(test_size=test_size, shuffle=False, stratify=None, random_state=random_state)
```
Comment (dsherry) [inline evalml/preprocessing/utils.py:68]: This is the change @freddyaboulton and I were discussion. Once this PR is merged we can update looking glass to use `split_data`

File: evalml/problem_types/utils.py [hunk 7]
```diff
@@ -49,3 +49,65 @@ def detect_problem_type(y):
         if (num_classes > 10):
             return ProblemTypes.REGRESSION
     return ProblemTypes.MULTICLASS
+
```
Comment (bchen1116) [inline evalml/problem_types/utils.py:52]: Should these new methods be tested?

**Review by bchen1116 — APPROVED** [review]
LGTM! Left a couple comments but I think you caught everything!

File: README.md [hunk 8]
```diff
@@ -34,8 +34,8 @@ X_train, X_test, y_train, y_test = evalml.preprocessing.split_data(X, y)
 #### Run AutoML
 ```python
 from evalml.automl import AutoMLSearch
-automl = AutoMLSearch(problem_type='binary')
-automl.search(X_train, y_train)
+automl = AutoMLSearch(X_train, y_train, problem_type='binary')
```
Comment (kmax12) [inline README.md:37]: nit pick, but it might be wise to make all these arguments named in our examples. that gives us more flexibility in the future when it comes to changing around positional argument and not breaking code. 

e.g in the future with unsupervised learning there might not be a `y` argument and the second positional argument could reference something else.

File: docs/source/start.ipynb [hunk 9]
```diff
@@ -61,32 +53,40 @@
...
+   "cell_type": "markdown",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "X_train, X_holdout, y_train, y_holdout = evalml.preprocessing.split_data(X, y, test_size=.2)"
+    "EvalML has many options to configure the pipeline search. At the minimum, we need to define an objective function. For simplicity, we will use the F1 score in this example. However, the real power of EvalML is in using domain-specific [objective functions](user_guide/objectives.ipynb) or [building your own](user_guide/objectives.ipynb).\n",
+    "\n",
+    "Below EvalML utilizes Bayesian optimization (EvalML's default optimizer) to search and find the best pipeline defined by the given objective.\n",
+    "\n",
+    "EvalML provides a number of parameters to control the search process. `max_batches` is one of the parameters which controls the stopping criterion for the AutoML search. It indicates the maximum number of rounds of AutoML to evaluate, where each round may train and score a variable number of pipelines. In this example, `max_batches` is set to 1.\n",
```
Comment (kmax12) [inline docs/source/start.ipynb:76]: 1. it might be better to move the explanation of `max_batches` until after the code example. it doesn't mean anything to see that parameter defined until it's actually used. for example, see in the featuretools docs how we explain the parameters to `entity_from_dataframe` at the end
image

2. not sure it's helpful to introduce the word "round" here since it's basically functioning as a synonym for batch without actually defining the word batch. 

3. We also don't give the user any intuition of when or why they'd want to use the parameter i.e it controls the "budget" evalml has to look for the best model. increasing the value allows evalml to search deeper for the best model, at the risk of overfitting.

Comment (kmax12) [reply docs/source/start.ipynb:76]: i see that this text wasn't actually change in the PR, so maybe best done as a follow up

File: evalml/problem_types/utils.py (hunk 7 above)
Comment (dsherry) [reply evalml/problem_types/utils.py:52]: Definitely :) will do

File: README.md (hunk 8 above)
Comment (dsherry) [reply README.md:37]: Good point. We already have `y` as a named arg in component fit/transform. Will do so here too.

File: docs/source/start.ipynb [hunk 10]
```diff
@@ -61,32 +53,40 @@
...
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "In order to validate the results of the pipeline creation and optimization process, we will save some of our data as a holdout set."
+    "__Note:__ To provide data to EvalML, it is recommended that you create a `DataTable` object using [the Woodwork project](https://woodwork.alteryx.com/en/stable/). Here, `split_data()` returns Woodwork data structures.\n",
+    "\n",
+    "EvalML also accepts and works well with pandas `DataFrames`. But using the `DataTable` makes it easy to control how EvalML will treat each feature, as a numeric feature, a categorical feature, a text feature or other type of feature. Woodwork's `DataTable` includes features like inferring when a categorical feature should be treated as a text feature. For this reason, if you don't provide Woodwork objects, EvalML will raise a warning."
```
[the remark this answers is not in the document]
Comment (dsherry) [reply docs/source/start.ipynb:65]: Thanks, done!
